source/common/upstream/load_balancer_impl.cc

```diff
@@ -615,41 +635,76 @@ bool LoadBalancerBase::isHostSetInPanic(const HostSet& host_set) const {
+    }
+  }
+  uint64_t total_upstream_hosts = 0;
+  for (const auto& locality_hosts : upstream_hosts_per_locality.get()) {
+    total_upstream_hosts += locality_hosts.size();
   }
 
-  // TODO(snowp): Should we ignore excluded hosts here too?
+  absl::FixedArray<LocalityPercentages> percentages(upstream_hosts_per_locality.get().size());
+  for (uint32_t i = 0; i < upstream_hosts_per_locality.get().size(); ++i) {
+    const auto& upstream_hosts = upstream_hosts_per_locality.get()[i];
+    if (upstream_hosts.empty()) {
+      // If there are no upstream hosts in a given locality, the upstream percentage is 0.
+      // We can't determine the locality of this group, so we can't find the corresponding local
+      // count. However, if there are no upstream hosts in a locality, the local percentage doesn't
+      // matter.
+      percentages[i] = LocalityPercentages{0, 0};
+      continue;
+    }
+    const auto& locality = upstream_hosts[0]->locality();
+
+    const auto& local_count_it = local_counts.find(locality);
+    const uint64_t local_count = local_count_it == local_counts.end() ? 0 : local_count_it->second;
 
-  size_t i = 0;
-  for (const auto& locality_hosts : hosts_per_locality.get()) {
-    ret[i++] = total_hosts > 0 ? 10000ULL * locality_hosts.size() / total_hosts : 0;
+    const uint64_t local_percentage =
+        total_local_hosts > 0 ? 10000ULL * local_count / total_local_hosts : 0;
+    const uint64_t upstream_percentage =
+        total_upstream_hosts > 0 ? 10000ULL * upstream_hosts.size() / total_upstream_hosts : 0;
+
+    percentages[i] = LocalityPercentages{local_percentage, upstream_percentage};
   }
+
+  return std::make_unique<absl::FixedArray<LocalityPercentages>>(std::move(percentages));
 }
 
 uint32_t ZoneAwareLoadBalancerBase::tryChooseLocalLocalityHosts(const HostSet& host_set) const {
   PerPriorityState& state = *per_priority_state_[host_set.priority()];
   ASSERT(state.locality_routing_state_ != LocalityRoutingState::NoLocalityRouting);
 
-  // At this point it's guaranteed to be at least 2 localities & local exists.
+  // At this point it's guaranteed to be at least 2 localities in the upstream host set.
   const size_t number_of_localities = host_set.healthyHostsPerLocality().get().size();
   ASSERT(number_of_localities >= 2U);
-  ASSERT(host_set.healthyHostsPerLocality().hasLocalLocality());
 
-  // Try to push all of the requests to the same locality first.
+  // Try to push all of the requests to the same locality if possible.
   if (state.locality_routing_state_ == LocalityRoutingState::LocalityDirect) {
+    ASSERT(host_set.healthyHostsPerLocality().hasLocalLocality());
     stats_.lb_zone_routing_all_directly_.inc();
     return 0;
   }
 
   ASSERT(state.locality_routing_state_ == LocalityRoutingState::LocalityResidual);
+  ASSERT(host_set.healthyHostsPerLocality().hasLocalLocality() ||
+         state.local_percent_to_route_ == 0);
```

Comment:
If state.local_percent_to_route_ > 0, then the upstream host set must have at least one host in the local locality.

Basically this is making sure we aren't accidentally trying to route locally when the host set doesn't have any hosts there, as that would have us routing unintentionally to locality 0 rather than the actual local locality.